### ISSUE TITLE
IPv4 fragmentation and reassembly

### ISSUE THREAD

**Issue by hannesm** [mirage/mirage-nat#31]
hello,

when working on fixing https://github.com/mirage/qubes-mirage-firewall/issues/73, I looked into the interaction between mirage-nat and qubes-mirage-firewall. If I understand correctly:
- every incoming piece of data is decoded into a `Nat_packet.t` (by `of_ipv4_packet`)
- every outgoing `Nat_packet.t` is encoded into a cstruct (by `into_cstruct`)

This works fine for complete IPv4 segments, but it fails for fragmented ones (on the receiving side, the transport header length exceeds the segment length (leading to an error message); on the sending side the buffer is smaller than the to-be-transmitted packet - leading to an error message).

(a) Since mirage-nat does encoding and decoding, I assume it should as well be responsible for handling fragmentation (otherwise, as seen in the branches [linked here](https://github.com/mirage/qubes-mirage-firewall/issues/73#issuecomment-515777076), qubes-mirage-firewall needs to decode, check fragmentation cache, and encode the full frame before passing it to mirage-nat). For doing this:
- `of_ipv4_packet` needs to take a `Fragments.Cache.t` (from mirage-tcpip/src/ipv4/fragments.ml), and return a `Fragments.Cache.t * t option`
- `into_cstruct` is a bit more involved: it needs the interface mtu to slice the data into multiple segments (and then either return a list of IPv4 segments (`to_cstruct`), or a way to get more buffers (`into_cstruct`))

(b) As alternative, mirage-nat could expect the first segment (which contains the transport header) to arrive first (since IP is not reliable, its fine to drop segments), decode (only) the transport header to do the NAT, and keep the IPv4 ID (which must be the same in all subsequent segments) in a table to NAT the subsequent segments. The upside is that no caches are necessary. The downside is that this only works if the MTU on the receive and send interface are the same.

NB: the `to_cstruct` / `into_cstruct` should preserve (and respect!) the `don't fragment` bit, and instead of fragmenting a segment send back an ICMP error.
NB: the branches linked above neither implement (a) or (b) (in the branches, qubes-mirage-firewall does the reassembly, the fragmentation is done by a weird interaction between q-m-f and mirage-nat), I'd prefer (a) (though (b) has its charme as well).

Tl;DR: What do you think, would you be happy with a PR that does (a) [and adapts your very fine example unikernel]?

**Comment by hannesm** [comment]
another approach could be to only work on hgih-level types (`Nat_packet.t`), and not provide any encoding and decoding functions.

**Comment by hannesm** [comment]
see #36 for an implementation PR
